fem/bilininteg_br2.cpp

```diff
@@ -0,0 +1,241 @@
+
+      R12 = 0.0;
+      R21 = 0.0;
+      R22 = 0.0;
+
+      Geometry::Type geom2 = Trans.Elem2->GetGeometryType();
+      factor = std::max(factor, double(Geometries.NumBdr(geom2)));
+   }
+   else
+   {
+      ndof2 = 0;
+   }
+
+   int ndofs = ndof1 + ndof2;
+
+   Re.SetSize(ndofs, ndofs);
+   MinvRe.SetSize(ndofs, ndofs);
+
+   elmat.SetSize(ndofs);
+   elmat = 0.0;
+
+   const IntegrationRule *ir = IntRule;
+   if (ir == NULL)
+   {
+      int order;
+      if (ndof2)
+      {
+         order = 2*std::max(el1.GetOrder(), el2.GetOrder());
+      }
+      else
+      {
+         order = 2*el1.GetOrder();
+      }
+      ir = &IntRules.Get(Trans.FaceGeom, order);
+   }
+
+   for (int p = 0; p < ir->GetNPoints(); p++)
+   {
+      const IntegrationPoint &ip = ir->IntPoint(p);
+      IntegrationPoint eip1, eip2;
+
+      Trans.Loc1.Transform(ip, eip1);
+      el1.CalcShape(eip1, shape1);
+      if (ndof2)
+      {
+         Trans.Loc2.Transform(ip, eip2);
+         el2.CalcShape(eip2, shape2);
+      }
+
+      double w = factor*sqrt(eta)*ip.weight*Trans.Face->Weight();
+      if (ndof2)
+      {
+         w /= 2;
+      }
+
+      for (int i = 0; i < ndof1; i++)
+      {
+         const double wsi = w*shape1(i);
+         for (int j = 0; j < ndof1; j++)
+         {
+            R11(i, j) += wsi*shape1(j);
+         }
+      }
+
+      if (ndof2)
+      {
+         for (int i = 0; i < ndof2; i++)
+         {
+            const double wsi = w*shape2(i);
+            for (int j = 0; j < ndof1; j++)
+            {
+               R21(i, j) += wsi*shape1(j);
+               R12(j, i) -= wsi*shape1(j);
+            }
+            for (int j = 0; j < ndof2; j++)
+            {
+               R22(i, j) -= wsi*shape2(j);
+            }
+         }
+      }
+   }
+
+   MinvR11 = R11;
+   M1inv.Solve(ndof1, ndof1, MinvR11.Data());
+   for (int i = 0; i < ndof1; i++)
+   {
+      for (int j = 0; j < ndof1; j++)
+      {
+         Re(i, j) = R11(i, j);
+         MinvRe(i, j) = MinvR11(i, j);
+      }
+   }
+
+   if (ndof2)
+   {
+      MinvR12 = R12;
+      MinvR21 = R21;
+      MinvR22 = R22;
+      M1inv.Solve(ndof1, ndof2, MinvR12.Data());
+      M2inv.Solve(ndof2, ndof1, MinvR21.Data());
+      M2inv.Solve(ndof2, ndof2, MinvR22.Data());
+
+      for (int i = 0; i < ndof2; i++)
+      {
+         for (int j = 0; j < ndof1; j++)
+         {
+            Re(ndof1 + i, j) = R21(i, j);
+            MinvRe(ndof1 + i, j) = MinvR21(i, j);
+
+            Re(j, ndof1 + i) = R12(j, i);
+            MinvRe(j, ndof1 + i) = MinvR12(j, i);
+         }
+         for (int j = 0; j < ndof2; j++)
+         {
+            Re(ndof1 + i, ndof1 + j) = R22(i, j);
+            MinvRe(ndof1 + i, ndof1 + j) = MinvR22(i, j);
+         }
+      }
+   }
+
+   // Compute the matrix associated with (r_e([u]), r_e([u])).
+   // The matrix for r_e([u]) is `MinvRe`, and so we need to form the product
+   // `MinvRe^T M MinvRe`. Canceling Minv and M, we obtain `Re^T M MinvRe`.
```

Comment:
I think the following correspond to the code bellow:
```suggestion
   // `MinvRe^T M MinvRe`. Canceling Minv and M, we obtain `Re^T MinvRe`.
```
However, why is it not:
```suggestion
   // `MinvRe^T M MinvRe`. Canceling Minv and M, we obtain `MinvRe^T Re`.
   MultAtB(MinvRe, Re, elmat);
```
?

Reply:
The notation is a bit confusing, `MinvRe^T` is supposed to mean `(MinvRe)^T`, so we have `(MinvRe)^T M MinvRe = Re^T Minv^T M Minv Re`, and `M` is symmetric, so we obtain `Re^T Minv Re`.